app/src/main/java/com/cmput301f20t21/bookfriends/ui/browse/BrowseViewModel.java

```diff
@@ -36,17 +48,67 @@ public BrowseViewModel(IAuthRepository authRepository, IRequestRepository reques
         this.requestRepository = requestRepository;
         this.bookRepository = bookRepository;
 
+        setupSearchedBooks(); // init the searched books before we fetch anything
         fetchBooks();
     }
 
-    public MutableLiveData<List<AvailableBook>> getBooks() {
-        return books;
+    /**
+     * create the searched books live data by making it listens to two source of changes:
+     * 1. the original books list: any changes of the books will trigger it to re-calculate results
+     * 2. the search keyword/query: changes of query should filter the searched books, too
+     */
+    private void setupSearchedBooks() {
+        searchedBooks.setValue(searchedBookData);
+        searchedBooks.addSource(searchQuery, keyword -> setSearchedBooksByKeywordAndBookData(keyword, bookData)); // keyword is new
+        searchedBooks.addSource(books, bookData -> setSearchedBooksByKeywordAndBookData(searchQuery.getValue(), bookData)); // bookData is new (this is a local one!)
+    }
+
+    /**
+     * filter the original bookData (or the updated bookData) with the keyword (or the updated one)
+     * and set that value to searchedBookData and its live data wrapper
+     *
+     * @param keyword
+     * @param bookData
+     */
+    private void setSearchedBooksByKeywordAndBookData(String keyword, List<AvailableBook> bookData) {
+        Function<String, Boolean> contains = ifNotNullAndContains(keyword);
+        List<AvailableBook> newBookData = bookData
+                .stream()
+                .filter(book -> contains.apply(book.getTitle()) ||
+                        contains.apply(book.getAuthor()) ||
+                        contains.apply(book.getOwner()))
```

Comment:
should we also allow search by ISBN?

Reply:
I think we should, it's actually pretty convenient.